concurrent/shared/src/test/scala/zio/concurrent/MVarSpec.scala

```diff
@@ -0,0 +1,58 @@
+package zio.concurrent
+
+import zio.test.Assertion._
+import zio.test._
+import zio._
+
+object MVarSpec extends ZIOSpecDefault {
+  val spec =
+    suite("MVarSpec")(
+      test("isEmpty") {
+        for {
+          v        <- MVar.empty[Unit]
+          v2       <- MVar.make(())
+          isEmpty1 <- v.isEmpty
+          _        <- v.put(())
+          isEmpty2 <- v.isEmpty
+          isEmpty3 <- v2.isEmpty
+          _        <- v2.take
+          isEmpty4 <- v2.isEmpty
+        } yield assert(isEmpty1)(isTrue) &&
+          assert(isEmpty2)(isFalse) &&
+          assert(isEmpty3)(isFalse) &&
+          assert(isEmpty4)(isTrue)
+      },
+      test("blocking put and take") {
+        for {
+          v   <- MVar.empty[Int]
+          _   <- ZIO.forkAll((1 to 100).map(i => v.put(i)))
+          res <- ZIO.foreach(1 to 100)(_ => v.take)
+        } yield assert(res)(hasSameElements(1 to 100))
+      },
+      test("non-blocking tryPut and tryTake") {
+        for {
+          v  <- MVar.empty[Unit]
+          r1 <- v.tryPut(())
+          r2 <- v.tryPut(())
+          r3 <- v.tryTake
+          r4 <- v.tryTake
+        } yield assert(r1)(isTrue) && assert(r2)(isFalse) && assert(r3)(isSome(isUnit)) && assert(r4)(isNone)
+      },
+      test("swap") {
+        for {
+          v  <- MVar.make(true)
+          r1 <- v.swap(false)
+          r2 <- v.take
+        } yield assert(r1)(isTrue) && assert(r2)(isFalse)
+      },
+      test("update") {
+        for {
+          v  <- MVar.empty[Boolean]
+          f  <- v.update(!_).fork
+          _  <- v.put(true)
+          _  <- f.join
```

Comment:
@adamgfraser This test was deadlocking (in CI only) when I put the `join` after the `v.take` below. It looks like under some condition, the main fiber is not yielding and the forked fiber above never gets the chance to do anything.

Reply:
@regiskuckaertz Is this issue that if the `join` is after `v.take` then we could have a situation where:

1. `v.put` is executed and puts `true` in the `MVar`
2. `v.take` is executed at takes `true` from the `MVar`
3. `f` attempts to update the `MVar` but since the `MVar` is now empty suspends indefinitely
4. We join `f` and suspend indefinitely

Reply:
🤦  yes you are correct